{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "58f70652-768c-4a58-9c4c-374ff940deb2",
   "metadata": {},
   "source": [
    "# **redoing the resue simulation sldiing window with another dna string**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "67c6815c-1dd3-406b-ad33-fde1e0c58ee1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from multiprocessing import Pool\n",
    "import random\n",
    "from random import randint\n",
    "from typing import List, Dict\n",
    "from numpy.random import choice\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy\n",
    "from Bio import SeqIO\n",
    "from tqdm import tqdm\n",
    "import copy \n",
    "import math \n",
    "import json\n",
    "import sys \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "95437ed6-0549-4ebb-bd7c-400d632f4228",
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulation_trial(cur_trial):\n",
    "\n",
    "    DNA = open(\"data/dif_DNA/2020_12_17_Trial10_DNA_final_cds_invariant50_9999.txt\").readlines()[0]\n",
    "    DNA_map = open(\"data/dif_DNA/2020_12_17_Trial10_DNA_map_cds_invariant50.txt\").readlines()[0] \n",
    "    #setting the number of generations \n",
    "    number_gens = len(DNA)*2\n",
    "    \n",
    "    # Description_of_sim,  Number_gens, where to i=find initial dna,   where to find initial dnamap \n",
    "\n",
    "    #MAKING STATIC OBJECTS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "\n",
    "    aa_conversion_dict = {\"TTT\": \"F\", \"TTC\": \"F\", \\\n",
    "    \"TTA\": \"L\", \"TTG\": \"L\", \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\", \\\n",
    "    \"ATT\": \"I\", \"ATC\": \"I\", \"ATA\": \"I\", \\\n",
    "    \"ATG\": \"M\", \\\n",
    "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\", \\\n",
    "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\", \"AGT\": \"S\", \"AGC\": \"S\", \\\n",
    "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\", \\\n",
    "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\", \\\n",
    "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\", \\\n",
    "    \"TAT\": \"Y\", \"TAC\": \"Y\", \\\n",
    "    \"CAT\": \"H\", \"CAC\": \"H\", \\\n",
    "    \"CAA\": \"Q\", \"CAG\": \"Q\", \\\n",
    "    \"AAT\": \"N\", \"AAC\": \"N\", \\\n",
    "    \"AAA\": \"K\", \"AAG\": \"K\", \\\n",
    "    \"GAT\": \"D\", \"GAC\": \"D\", \\\n",
    "    \"GAA\": \"E\", \"GAG\": \"E\", \\\n",
    "    \"TGT\": \"C\", \"TGC\": \"C\", \\\n",
    "    \"TGG\": \"W\", \\\n",
    "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\", \\\n",
    "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\", \\\n",
    "    \"TAA\": \"*\", \"TAG\": \"*\", \"TGA\": \"*\"} \\\n",
    "\n",
    "    #FUNCTIONS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "\n",
    "    def reverse_complement(dna):\n",
    "        \"\"\"\n",
    "        Str --> str \n",
    "        Note that this function will revComp everything: need appropruate if conditional to make sure you are 64-->32 not 64--> complementary 64 \n",
    "        \"\"\"\n",
    "        complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "        return \"fuckup: used the revcomp\"\n",
    "\n",
    "    def base_to_mutate(DNA, current_weights): \n",
    "        \"\"\"\n",
    "        (DNA: str, weights: Dict) -> int\n",
    "\n",
    "        Will return the index of a random choice of a base that will be mutated based on probabilities given by the\n",
    "        weights dictionary (i.e. the model)\n",
    "        ex. \"ATCGTA\" --> index 3 (\"G\") will mutate\n",
    "        \"\"\"\n",
    "        #DNA STR --> WEIGHTS LIST INDEXING DOUBLE CHECK \n",
    "        indices = []\n",
    "        indices = list(i for i in range(1, len(DNA)-1))\n",
    "        if (len(indices)) != (len(current_weights)): \n",
    "            print(\"ERROR, FUCKED UP\")\n",
    "            print(\"length of indices = \" + str(len(indices)))\n",
    "            print(\"length of weights = \" + str(len(current_weights)))\n",
    "\n",
    "        # NORMALIZE THE POPULATION OF WEIGHTS    \n",
    "        total_freq = sum(current_weights) \n",
    "        normalized_weights = copy.copy(current_weights)\n",
    "        for index, value in enumerate(current_weights):\n",
    "            normalized_weights[index] = value/total_freq \n",
    "\n",
    "        # DRAW THE INDEX OF THE BASE THAT WILL BE MUTATED\n",
    "        base_index = choice(indices, p=normalized_weights) \n",
    "\n",
    "        # RETURN THE INDEX\n",
    "        return base_index\n",
    "\n",
    "    def current_frequencies(DNA, weights, index, current_weights): #and by weights, she means the \"model\" dictionary \n",
    "        \"\"\"\n",
    "        (DNA: str, weights list, index = chosen site) -> updated weights List \n",
    "        only chnages the 3 bases around the index chosen \n",
    "        \"\"\"\n",
    "        #CREATE THE 3 SOURROUNDING TRIPLETS\n",
    "        triplet_preceding = DNA[index-2:index+1] #dont worry this indexing is double checked \n",
    "        triplet_center = DNA[index-1:index+2]\n",
    "        triplet_following = DNA[index:index+3]\n",
    "        triplet_list = [triplet_preceding, triplet_center, triplet_following]\n",
    "\n",
    "        #64-->32 \n",
    "        for triplet in triplet_list: \n",
    "            if triplet not in weights: \n",
    "                i = triplet_list.index(triplet)\n",
    "                triplet_list[i] = reverse_complement(triplet)\n",
    "\n",
    "        #GET THE WEIGHTS FOR EACH TRIPLET AND UPDATE WEIGHTS (LIST) OBJECT \n",
    "        current_weights[index-2]= weights[triplet_list[0]][0]    #note that it's the expectedindex -1 as the index = position in string (from 1 to n-1) but here the items are the codons ...? \n",
    "        current_weights[index-1]= weights[triplet_list[1]][0]\n",
    "        current_weights[index] = weights[triplet_list[2]][0]\n",
    "\n",
    "        return current_weights\n",
    "\n",
    "    def base_to_turn_into(i, model, DNA):\n",
    "        \"\"\"\n",
    "        (i: int, weights: Dict, DNA: str) -> str (the base A/T/C/G)\n",
    "        \"\"\"\n",
    "\n",
    "        # GRAB THE TRIPLET AND COMPLEMENT IT IF NEED BE.\n",
    "        triplet = DNA[i - 1:i+2] \n",
    "        if triplet not in model:\n",
    "                triplet = reverse_complement(triplet)\n",
    "\n",
    "        # GENERATE A LIST OF THE TRIPLETS AND WEIGHTS OF THE THREE TRIPLETS THE CHOSEN TRIPLET CAN TURN INTO \n",
    "        triplets = list(key for key in model[triplet][1].keys())# ex. \"ATC\" --> [\"AAC\", \"AGC\", \"ACC\"] (triplets)\n",
    "        current_weights_to_turn_into = list(value for value in model[triplet][1].values()) # ex. \"ATCGTA\" --> [0.3, 0.5, 0.1, 0.1] (weights of triplet base changes) (these are already normalized!)\n",
    "\n",
    "        # MAKE AND RETURN THE CHOICE OF BASE TO TURN INTO.\n",
    "        return choice(triplets, p = current_weights_to_turn_into)[1] # <-- the [1] will select the central base (middle N base out of strign 3)\n",
    "\n",
    "    def codingYesNo(ab_index, mb, DNA, DNA_map):\n",
    "        \"\"\"\n",
    "        (ab_index: int, mb = base A/T/C/G, DNA: str, DNA_map: List) -> List or None\n",
    "\n",
    "        Will check if the ancestral base is part of the coding/ non-coding/invariant region.\n",
    "        If part of a noncoding region, it will return \"None\".\n",
    "        If part of a coding region, it will return the ancestral and mutated codons in a list with two entries [ancestral_codon, mutated_codon] \n",
    "        If part of an invariant region, returns int(3)\n",
    "        \"\"\"\n",
    "        if DNA_map[ab_index] == \"N\": # if it's not non-coding (then it's coding)...\n",
    "            return None\n",
    "        elif DNA_map[ab_index] == \"3\": #if invariant \n",
    "            return 3\n",
    "        else:\n",
    "            if DNA_map[ab_index] == \"0\": # if we're at the first position...\n",
    "                codon = DNA[ab_index] + DNA[ab_index + 1] + DNA[ab_index + 2] # take the 012 codon\n",
    "                m_codon = mb + DNA[ab_index + 1] + DNA[ab_index + 2] # make the mutated codon\n",
    "            elif DNA_map[ab_index] == \"1\": # if we're at the second position...\n",
    "                codon = DNA[ab_index - 1] + DNA[ab_index] + DNA[ab_index + 1] # take the 012 codon\n",
    "                m_codon = DNA[ab_index - 1] + mb + DNA[ab_index + 1] # make the mutated codon\n",
    "            elif DNA_map[ab_index] == \"2\": # if we're at the third position...\n",
    "                codon = DNA[ab_index - 2] + DNA[ab_index - 1] + DNA[ab_index] # take the 012 codon\n",
    "                m_codon = DNA[ab_index - 2] + DNA[ab_index - 1] + mb # make the mutated codon\n",
    "\n",
    "            return [codon, m_codon] # if it's coding, return the ancestral and mutated codons\n",
    "\n",
    "    def blosum(a_m_codons, codon_aa_conversionDict, blosum_dict):\n",
    "        \"\"\"\n",
    "        (a_m_codons: List [ancestral_codon, mutated_codon], codon_aa_conversionDict = triplet-->aa dict, blosum_dict = blosum matrix ) -> bool (True/False     \n",
    "        Will return True if the change is acceptable and False if the change is not acceptable.  \n",
    "        ex. [\"CGT\", \"CAT\"] --> R to H is accepted at 0.6, or 60%. If the function accepts it, it will return True.\n",
    "        \"\"\"\n",
    "\n",
    "        a_aa = codon_aa_conversionDict[a_m_codons[0]] #ancestral a.a \n",
    "        m_aa = codon_aa_conversionDict[a_m_codons[1]]  #mutated a.a \n",
    "\n",
    "        return choice([True, False], p = [blosum_dict[a_aa][m_aa], (1-(blosum_dict[a_aa][m_aa]))]) \n",
    "\n",
    "    def mutate_chromosome(DNA, ab_index, mb):\n",
    "        \"\"\"\n",
    "        (DNA:str, ab_index:number, site chosen, mb:str) -> str (mutated DNA)\n",
    "        Will return a new DNA string with a random base mutated to a new base as given by the previous functions.\n",
    "        \"\"\"\n",
    "        return DNA[:ab_index] + mb + DNA[ab_index + 1:]\n",
    "\n",
    "    def average_mutability(current_weights):\n",
    "        \"\"\"\n",
    "        (weights: list) -> float\n",
    "        needs to come before the current_frequencies function \n",
    "        Will return a single mutability score which is the average of the triplet mutation frequencies in the string.\n",
    "        ex. \"ATCGTA\" --> [0.02, 0.31, 0.15, 0.07] --> (0.02+0.31+0.15+0.07)/4 --> 0.1375\n",
    "        \"\"\"\n",
    "        mean = np.mean(current_weights)\n",
    "        return mean # return the average (number) \n",
    "\n",
    "    def exon_mutability(DNA_map,current_weights):\n",
    "        \"\"\"\n",
    "        (DNA_map = str, current_weights = list) -> float\n",
    "        Same function as average_mutability, but for CODING REGIONs ONLY (why you need the map) \n",
    "        \"\"\"\n",
    "        #SUM MUTABILITY ACROSS THE ENTIRE DNA MAP STRING USING DNA_MAP IF CONDITIONAL \n",
    "        mutability = 0 \n",
    "        count = 0 \n",
    "        for i in range(1, len(DNA_map)-1): # this will exclude the first base and the last base\n",
    "            if DNA_map[i] != \"N\": # if we are in the coding region\n",
    "                mutability += current_weights[i-1]# add the mutability of the coding triplet to the total --> and YES ive chedked this, dont freak out the indices match up... \n",
    "                count += 1             \n",
    "        return mutability/count \n",
    "\n",
    "    def intron_mutability(DNA_map, current_weights):\n",
    "        \"\"\"\n",
    "        (DNA_map = str, current_weights = list) -> float\n",
    "        Same function as average_mutability, but for nonCODING REGIONs ONLY (why you need the map)\n",
    "        \"\"\"\n",
    "        #SUM MUTABILITY ACROSS THE ENTIRE DNA MAP STRING USING DNA_MAP IF CONDITIONAL \n",
    "        mutability = 0 \n",
    "        count = 0\n",
    "        for i in range(1, len(DNA_map)-1): # this will exclude the first base and the last base\n",
    "            if DNA_map[i] == \"N\": # if we are in the coding region\n",
    "                mutability += current_weights[i-1]# add the mutability of the coding triplet to the total --> and YES ive chedked this, dont freak out the indices match up... \n",
    "                count += 1\n",
    "        return mutability/count \n",
    "\n",
    "    #MAKING VARIABLE OBJECTS \n",
    "\n",
    "    #dictionary of CDS greater than a certain size (here >4k) \n",
    "    cds_file_obj = open(\"../Human_exon_data/Homo_sapiens.GRCh38.cds.all.fa\")\n",
    "    total_cds_dict = {}\n",
    "    for gene in SeqIO.parse(cds_file_obj, \"fasta\"): \n",
    "        if len(gene.seq) >= 4000: \n",
    "            cds_map = \"012\"*int(len(gene.seq)/3)\n",
    "            if len(cds_map) == len(gene.seq) and \"N\" not in gene.seq: \n",
    "                total_cds_dict[gene.id] = [str(gene.seq), cds_map]\n",
    "    print(\"total number of CDS where len >= 4k is \"+str(len(total_cds_dict)))                                #! print statement? \n",
    "\n",
    "    #blosum matrix \n",
    "    blosum_dict = {'T': {'T': 1.0, 'Q': 0.5, 'C': 0.4, 'M': 0.5, 'A': 0.6, 'B': 0.5, 'K': 0.5, 'E': 0.5, 'P': 0.4, 'R': 0.4, 'G': 0.3, 'J': 0.4, 'F': 0.3, 'X': 0.5, 'Z': 0.5, '*': 0.0, 'N': 0.6, 'V': 0.5, 'H': 0.4, 'W': 0.2, 'L': 0.4, 'S': 0.7, 'D': 0.4, 'I': 0.5, 'Y': 0.4}, \\\n",
    "     'Q': {'T': 0.5, 'Q': 1.0, 'C': 0.2, 'M': 0.6, 'A': 0.5, 'B': 0.5, 'K': 0.7, 'E': 0.8, 'P': 0.4, 'R': 0.7, 'G': 0.3, 'J': 0.3, 'F': 0.2, 'X': 0.5, 'Z': 1.0, '*': 0.0, 'N': 0.6, 'V': 0.3, 'H': 0.7, 'W': 0.3, 'L': 0.3, 'S': 0.5, 'D': 0.5, 'I': 0.2, 'Y': 0.3}, \\\n",
    "     'C': {'T': 0.4, 'Q': 0.2, 'C': 1.0, 'M': 0.4, 'A': 0.5, 'B': 0.2, 'K': 0.2, 'E': 0.0, 'P': 0.2, 'R': 0.1, 'G': 0.2, 'J': 0.4, 'F': 0.3, 'X': 0.5, 'Z': 0.1, '*': 0.0, 'N': 0.2, 'V': 0.4, 'H': 0.1, 'W': 0.2, 'L': 0.4, 'S': 0.4, 'D': 0.1, 'I': 0.4, 'Y': 0.2}, \\\n",
    "     'M': {'T': 0.5, 'Q': 0.6, 'C': 0.4, 'M': 1.0, 'A': 0.4, 'B': 0.2, 'K': 0.4, 'E': 0.3, 'P': 0.3, 'R': 0.4, 'G': 0.2, 'J': 0.8, 'F': 0.5, 'X': 0.5, 'Z': 0.4, '*': 0.0, 'N': 0.3, 'V': 0.6, 'H': 0.3, 'W': 0.4, 'L': 0.8, 'S': 0.4, 'D': 0.2, 'I': 0.7, 'Y': 0.4}, \\\n",
    "     'A': {'T': 0.6, 'Q': 0.5, 'C': 0.5, 'M': 0.4, 'A': 1.0, 'B': 0.4, 'K': 0.5, 'E': 0.5, 'P': 0.5, 'R': 0.4, 'G': 0.6, 'J': 0.4, 'F': 0.3, 'X': 0.5, 'Z': 0.5, '*': 0.0, 'N': 0.4, 'V': 0.5, 'H': 0.4, 'W': 0.2, 'L': 0.4, 'S': 0.7, 'D': 0.3, 'I': 0.4, 'Y': 0.3}, \\\n",
    "     'B': {'T': 0.5, 'Q': 0.5, 'C': 0.2, 'M': 0.2, 'A': 0.4, 'B': 1.0, 'K': 0.5, 'E': 0.7, 'P': 0.3, 'R': 0.4, 'G': 0.4, 'J': 0.1, 'F': 0.2, 'X': 0.5, 'Z': 0.6, '*': 0.0, 'N': 1.0, 'V': 0.2, 'H': 0.5, 'W': 0.0, 'L': 0.1, 'S': 0.6, 'D': 1.0, 'I': 0.1, 'Y': 0.2}, \\\n",
    "     'K': {'T': 0.5, 'Q': 0.7, 'C': 0.2, 'M': 0.4, 'A': 0.5, 'B': 0.5, 'K': 1.0, 'E': 0.6, 'P': 0.4, 'R': 0.8, 'G': 0.4, 'J': 0.3, 'F': 0.2, 'X': 0.5, 'Z': 0.7, '*': 0.0, 'N': 0.6, 'V': 0.3, 'H': 0.5, 'W': 0.1, 'L': 0.3, 'S': 0.5, 'D': 0.5, 'I': 0.2, 'Y': 0.3}, \\\n",
    "     'E': {'T': 0.5, 'Q': 0.8, 'C': 0.0, 'M': 0.3, 'A': 0.5, 'B': 0.7, 'K': 0.6, 'E': 1.0, 'P': 0.4, 'R': 0.5, 'G': 0.3, 'J': 0.2, 'F': 0.1, 'X': 0.5, 'Z': 1.0, '*': 0.0, 'N': 0.5, 'V': 0.3, 'H': 0.5, 'W': 0.1, 'L': 0.2, 'S': 0.5, 'D': 0.7, 'I': 0.2, 'Y': 0.2}, \\\n",
    "     'P': {'T': 0.4, 'Q': 0.4, 'C': 0.2, 'M': 0.3, 'A': 0.5, 'B': 0.3, 'K': 0.4, 'E': 0.4, 'P': 1.0, 'R': 0.3, 'G': 0.3, 'J': 0.2, 'F': 0.2, 'X': 0.5, 'Z': 0.4, '*': 0.0, 'N': 0.3, 'V': 0.3, 'H': 0.3, 'W': 0.1, 'L': 0.2, 'S': 0.4, 'D': 0.3, 'I': 0.2, 'Y': 0.2}, \\\n",
    "     'R': {'T': 0.4, 'Q': 0.7, 'C': 0.1, 'M': 0.4, 'A': 0.4, 'B': 0.4, 'K': 0.8, 'E': 0.5, 'P': 0.3, 'R': 1.0, 'G': 0.3, 'J': 0.3, 'F': 0.2, 'X': 0.5, 'Z': 0.6, '*': 0.0, 'N': 0.5, 'V': 0.3, 'H': 0.6, 'W': 0.2, 'L': 0.3, 'S': 0.5, 'D': 0.3, 'I': 0.2, 'Y': 0.3}, \\\n",
    "     'G': {'T': 0.3, 'Q': 0.3, 'C': 0.2, 'M': 0.2, 'A': 0.6, 'B': 0.4, 'K': 0.4, 'E': 0.3, 'P': 0.3, 'R': 0.3, 'G': 1.0, 'J': 0.1, 'F': 0.1, 'X': 0.5, 'Z': 0.3, '*': 0.0, 'N': 0.5, 'V': 0.1, 'H': 0.3, 'W': 0.2, 'L': 0.1, 'S': 0.5, 'D': 0.4, 'I': 0.1, 'Y': 0.1}, \\\n",
    "     'J': {'T': 0.4, 'Q': 0.3, 'C': 0.4, 'M': 0.8, 'A': 0.4, 'B': 0.1, 'K': 0.3, 'E': 0.2, 'P': 0.2, 'R': 0.3, 'G': 0.1, 'J': 1.0, 'F': 0.6, 'X': 0.5, 'Z': 0.2, '*': 0.0, 'N': 0.2, 'V': 0.7, 'H': 0.2, 'W': 0.3, 'L': 1.0, 'S': 0.3, 'D': 0.1, 'I': 0.9, 'Y': 0.4}, \\\n",
    "     'F': {'T': 0.3, 'Q': 0.2, 'C': 0.3, 'M': 0.5, 'A': 0.3, 'B': 0.2, 'K': 0.2, 'E': 0.1, 'P': 0.2, 'R': 0.2, 'G': 0.1, 'J': 0.6, 'F': 1.0, 'X': 0.5, 'Z': 0.2, '*': 0.0, 'N': 0.2, 'V': 0.4, 'H': 0.4, 'W': 0.6, 'L': 0.6, 'S': 0.3, 'D': 0.1, 'I': 0.5, 'Y': 0.9}, \\\n",
    "     'X': {'T': 0.5, 'Q': 0.5, 'C': 0.5, 'M': 0.5, 'A': 0.5, 'B': 0.5, 'K': 0.5, 'E': 0.5, 'P': 0.5, 'R': 0.5, 'G': 0.5, 'J': 0.5, 'F': 0.5, 'X': 0.5, 'Z': 0.5, '*': 0.0, 'N': 0.5, 'V': 0.5, 'H': 0.5, 'W': 0.5, 'L': 0.5, 'S': 0.5, 'D': 0.5, 'I': 0.5, 'Y': 0.5}, \\\n",
    "     'Z': {'T': 0.5, 'Q': 1.0, 'C': 0.1, 'M': 0.4, 'A': 0.5, 'B': 0.6, 'K': 0.7, 'E': 1.0, 'P': 0.4, 'R': 0.6, 'G': 0.3, 'J': 0.2, 'F': 0.2, 'X': 0.5, 'Z': 1.0, '*': 0.0, 'N': 0.5, 'V': 0.3, 'H': 0.6, 'W': 0.2, 'L': 0.2, 'S': 0.5, 'D': 0.7, 'I': 0.2, 'Y': 0.3}, \\\n",
    "     '*': {'T': 0.0, 'Q': 0.0, 'C': 0.0, 'M': 0.0, 'A': 0.0, 'B': 0.0, 'K': 0.0, 'E': 0.0, 'P': 0.0, 'R': 0.0, 'G': 0.0, 'J': 0.0, 'F': 0.0, 'X': 0.0, 'Z': 0.0, '*': 0.7, 'N': 0.0, 'V': 0.0, 'H': 0.0, 'W': 0.0, 'L': 0.0, 'S': 0.0, 'D': 0.0, 'I': 0.0, 'Y': 0.0}, \\\n",
    "     'N': {'T': 0.6, 'Q': 0.6, 'C': 0.2, 'M': 0.3, 'A': 0.4, 'B': 1.0, 'K': 0.6, 'E': 0.5, 'P': 0.3, 'R': 0.5, 'G': 0.5, 'J': 0.2, 'F': 0.2, 'X': 0.5, 'Z': 0.5, '*': 0.0, 'N': 1.0, 'V': 0.2, 'H': 0.6, 'W': 0.1, 'L': 0.2, 'S': 0.6, 'D': 0.7, 'I': 0.2, 'Y': 0.3}, \\\n",
    "     'V': {'T': 0.5, 'Q': 0.3, 'C': 0.4, 'M': 0.6, 'A': 0.5, 'B': 0.2, 'K': 0.3, 'E': 0.3, 'P': 0.3, 'R': 0.3, 'G': 0.1, 'J': 0.7, 'F': 0.4, 'X': 0.5, 'Z': 0.3, '*': 0.0, 'N': 0.2, 'V': 1.0, 'H': 0.2, 'W': 0.3, 'L': 0.6, 'S': 0.4, 'D': 0.1, 'I': 0.9, 'Y': 0.3}, \\\n",
    "     'H': {'T': 0.4, 'Q': 0.7, 'C': 0.1, 'M': 0.3, 'A': 0.4, 'B': 0.5, 'K': 0.5, 'E': 0.5, 'P': 0.3, 'R': 0.6, 'G': 0.3, 'J': 0.2, 'F': 0.4, 'X': 0.5, 'Z': 0.6, '*': 0.0, 'N': 0.6, 'V': 0.2, 'H': 1.0, 'W': 0.3, 'L': 0.2, 'S': 0.4, 'D': 0.4, 'I': 0.2, 'Y': 0.7}, \\\n",
    "     'W': {'T': 0.2, 'Q': 0.3, 'C': 0.2, 'M': 0.4, 'A': 0.2, 'B': 0.0, 'K': 0.1, 'E': 0.1, 'P': 0.1, 'R': 0.2, 'G': 0.2, 'J': 0.3, 'F': 0.6, 'X': 0.5, 'Z': 0.2, '*': 0.0, 'N': 0.1, 'V': 0.3, 'H': 0.3, 'W': 1.0, 'L': 0.3, 'S': 0.2, 'D': 0.0, 'I': 0.2, 'Y': 0.8}, \\\n",
    "     'L': {'T': 0.4, 'Q': 0.3, 'C': 0.4, 'M': 0.8, 'A': 0.4, 'B': 0.1, 'K': 0.3, 'E': 0.2, 'P': 0.2, 'R': 0.3, 'G': 0.1, 'J': 1.0, 'F': 0.6, 'X': 0.5, 'Z': 0.2, '*': 0.0, 'N': 0.2, 'V': 0.6, 'H': 0.2, 'W': 0.3, 'L': 1.0, 'S': 0.3, 'D': 0.1, 'I': 0.7, 'Y': 0.4}, \\\n",
    "     'S': {'T': 0.7, 'Q': 0.5, 'C': 0.4, 'M': 0.4, 'A': 0.7, 'B': 0.6, 'K': 0.5, 'E': 0.5, 'P': 0.4, 'R': 0.5, 'G': 0.5, 'J': 0.3, 'F': 0.3, 'X': 0.5, 'Z': 0.5, '*': 0.0, 'N': 0.6, 'V': 0.4, 'H': 0.4, 'W': 0.2, 'L': 0.3, 'S': 1.0, 'D': 0.5, 'I': 0.3, 'Y': 0.3}, \\\n",
    "     'D': {'T': 0.4, 'Q': 0.5, 'C': 0.1, 'M': 0.2, 'A': 0.3, 'B': 1.0, 'K': 0.5, 'E': 0.7, 'P': 0.3, 'R': 0.3, 'G': 0.4, 'J': 0.1, 'F': 0.1, 'X': 0.5, 'Z': 0.7, '*': 0.0, 'N': 0.7, 'V': 0.1, 'H': 0.4, 'W': 0.0, 'L': 0.1, 'S': 0.5, 'D': 1.0, 'I': 0.1, 'Y': 0.2}, \\\n",
    "     'I': {'T': 0.5, 'Q': 0.2, 'C': 0.4, 'M': 0.7, 'A': 0.4, 'B': 0.1, 'K': 0.2, 'E': 0.2, 'P': 0.2, 'R': 0.2, 'G': 0.1, 'J': 0.9, 'F': 0.5, 'X': 0.5, 'Z': 0.2, '*': 0.0, 'N': 0.2, 'V': 0.9, 'H': 0.2, 'W': 0.2, 'L': 0.7, 'S': 0.3, 'D': 0.1, 'I': 1.0, 'Y': 0.4}, \\\n",
    "     'Y': {'T': 0.4, 'Q': 0.3, 'C': 0.2, 'M': 0.4, 'A': 0.3, 'B': 0.2, 'K': 0.3, 'E': 0.2, 'P': 0.2, 'R': 0.3, 'G': 0.1, 'J': 0.4, 'F': 0.9, 'X': 0.5, 'Z': 0.3, '*': 0.0, 'N': 0.3, 'V': 0.3, 'H': 0.7, 'W': 0.8, 'L': 0.4, 'S': 0.3, 'D': 0.2, 'I': 0.4, 'Y': 1.0}}\n",
    "\n",
    "    #mutability model \n",
    "    model = json.load(open(\"../Human_mutability_model/Model_2020_12_02_genomeWide.txt\"))\n",
    "\n",
    "\n",
    "\n",
    "    #READING IN THE EXON DICTIONARIES \n",
    "    #file_exonsInsertPos = sys.argv[4]\n",
    "    #exons_insertion_position_dict = json.loads(open(file_exonsInsertPos).readlines()[0])\n",
    "    #file_exonsToUse = sys.argv[5]\n",
    "    #exons_to_use_dict = json.loads(open(file_exonsToUse).readlines()[0])\n",
    "\n",
    "    #creating the list of initial weights \n",
    "    weights = []\n",
    "    for i in range(1,len(DNA)-1): # dont want it to run all the way to the end: need to -1 so last centre index is second last in DNA string \n",
    "        if (DNA[i-1:i+2]) in model: \n",
    "            weights.append(model[(DNA[i-1:i+2])][0])\n",
    "        else: \n",
    "            weights.append(model[reverse_complement(DNA[i-1:i+2])][0])\n",
    "\n",
    "    #writing initial conditions to file \n",
    "    #NameForThisSim = sys.argv[1]\n",
    "    #text_file = open(\"Scripting_data_trial2/{t}_DNA_gen0_cds.txt\".format(t=NameForThisSim), \"w\")                                #! need to change file creation: use the {}.format... file structure? \n",
    "    #n = text_file.write(DNA)\n",
    "    #text_file.close()\n",
    "    #text_file = open(\"Scripting_data_trial2/{t}_DNA_map_cds_invariant50.txt\".format(t=NameForThisSim), \"w\")\n",
    "    #n = text_file.write(DNA_map)\n",
    "    #text_file.close()\n",
    "    #text_file = open(\"Scripting_data_trial2/{t}_exon_insertion_dict.txt\".format(t=NameForThisSim), \"w\")\n",
    "    #n = text_file.write(json.dumps(exons_insertion_position_dict))\n",
    "    #text_file.close()\n",
    "    #text_file = open(\"Scripting_data_trial2/{t}_exons_toUseSeqeunce_dict.txt\".format(t=NameForThisSim), \"w\")\n",
    "    #n = text_file.write(json.dumps(exons_to_use_dict))\n",
    "    #text_file.close()\n",
    "\n",
    "    #SIMULATION ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "\n",
    "    #creating the black plotting lists \n",
    "    av_mutability = []\n",
    "    intron_mut = []\n",
    "    exon_mut = []\n",
    "\n",
    "    #creating the list for num times to print \n",
    "    gens_for_output = []\n",
    "\n",
    "    for x in range(0,int(number_gens)+1, 100000): \n",
    "        gens_for_output.append(x)\n",
    "\n",
    "    for x in tqdm(range(int(number_gens)+1)):\n",
    "\n",
    "        #selecting the base to mutate \n",
    "        ab_index = base_to_mutate(DNA, weights) # ex. \"mutate position 25 (\"G\")\"   \n",
    "\n",
    "        if len(DNA)-ab_index >= 3 and ab_index >= 3 : #making sure dont pick the ends \n",
    "\n",
    "            mb = base_to_turn_into(ab_index, model, DNA) # ex. \"position 25 (\"G\") will become \"A\"    \n",
    "\n",
    "            #determining if alloweed to mutate. if so, then mutating \n",
    "            if codingYesNo(ab_index, mb, DNA, DNA_map) == None:\n",
    "                mDNA = mutate_chromosome(DNA, ab_index, mb) # always mutate \n",
    "                DNA = mDNA # the mutated DNA is the new DNA \n",
    "            elif codingYesNo(ab_index, mb, DNA, DNA_map) != None and codingYesNo(ab_index, mb, DNA, DNA_map) != 3: # if the mutation is in a CODING REGION...\n",
    "                a_m_codons = codingYesNo(ab_index, mb, DNA, DNA_map) # check to see if the change is acceptable\n",
    "                if blosum(a_m_codons, aa_conversion_dict, blosum_dict) == True: # if the change is acceptable...\n",
    "                    mDNA = mutate_chromosome(DNA, ab_index, mb) # mutate it (otherwise do nothing)\n",
    "                    DNA = mDNA # the mutated DNA is the new DNA    \n",
    "\n",
    "            #updating the weights lits \n",
    "            weights = current_frequencies(DNA, model, ab_index, weights)\n",
    "\n",
    "            #updating the mutability values \n",
    "            #av_mutability.append(average_mutability(weights)) # AVERAGE MUTABILITY OF WHOLE STRING\n",
    "            #exon_mut.append(exon_mutability(DNA_map, weights)) # AVEARAGE MUTABILITY OF EXON (DNA_map, index, current_weights, total_intron_length)\n",
    "            #intron_mut.append(intron_mutability(DNA_map, weights)) # AVERAGE MUTABILITY OF INTRON  \n",
    "\n",
    "        if x in gens_for_output: \n",
    "            # WRITING DATA TO FILES \n",
    "            NameForThisSim =\"redoTrial\"+str(cur_trial)\n",
    "\n",
    "    #         file = open(\"Scripting_data_trial2/Nonsyn_output/{t}_avMutability_cds_invariant50_{g}.txt\".format(t=NameForThisSim, g = x), \"w+\")\n",
    "    #         file.write(str(av_mutability))\n",
    "    #         file.close()\n",
    "\n",
    "    #         file = open(\"Scripting_data_trial2/Nonsyn_output/{t}_intronMut_cds_invariant50_{g}.txt\".format(t=NameForThisSim, g = x), \"w+\")\n",
    "    #         file.write(str(intron_mut))\n",
    "    #         file.close()\n",
    "\n",
    "    #         file = open(\"Scripting_data_trial2/Nonsyn_output/{t}_exonMut_cds_invariant50_{g}.txt\".format(t=NameForThisSim, g = x), \"w+\")\n",
    "    #         file.write(str(exon_mut))\n",
    "    #         file.close()\n",
    "\n",
    "            file = open(\"data/same_DNA/{t}_DNA_final_cds_invariant50_{g}.txt\".format(t=NameForThisSim, g = x), \"w+\")\n",
    "            file.write(DNA)\n",
    "            file.close()\n",
    "\n",
    "        file = open(\"data/same_DNA/{t}_DNA_final_cds_invariant50_{g}.txt\".format(t=NameForThisSim, g = x), \"w+\")\n",
    "        file.write(DNA)\n",
    "        file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8089e27b-7269-449f-9dbe-57844c4dddbc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 27%|██████████████████████████████▌                                                                                 | 3/11 [00:00<00:00, 26.23it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of CDS where len >= 4k is 3473\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 11/11 [00:00<00:00, 25.76it/s]\n"
     ]
    }
   ],
   "source": [
    "simulation_trial(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08ccbe58-be4e-4869-8cc4-4a771656f653",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of CDS where len >= 4k is 3473\n",
      "total number of CDS where len >= 4k is 3473\n",
      "total number of CDS where len >= 4k is 3473total number of CDS where len >= 4k is 3473\n",
      "\n",
      "total number of CDS where len >= 4k is 3473\n",
      "total number of CDS where len >= 4k is 3473\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                                                                    | 0/209179 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of CDS where len >= 4k is 3473\n",
      "total number of CDS where len >= 4k is 3473"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                                                                    | 0/209179 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "total number of CDS where len >= 4k is 3473\n",
      "total number of CDS where len >= 4k is 3473\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 93%|██████████████████████████████████████████████████████████████████████████████████████████████       | 194800/209179 [2:00:34<08:59, 26.65it/s]"
     ]
    }
   ],
   "source": [
    "if __name__ == '__main__':\n",
    "    with Pool(10) as p:\n",
    "        print(p.map(simulation_trial, [1, 2, 3,4,5,6,7,8,9,10]))\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e87e7963-5152-4116-9496-6343c56e9775",
   "metadata": {},
   "outputs": [],
   "source": [
    "2+2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-mutEq]",
   "language": "python",
   "name": "conda-env-.conda-mutEq-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
